### PR TITLE
Add health check package with watchdogs

### PR DESCRIPTION
**Pull request by AndreyTretyak**
Add package with watchdogs

### CODE REVIEW

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks
+{
+	internal class HttpEndpointHealthCheck : AbstractHealthCheck
+	{
+		private const string Host = "localhost";
+
+		public static string HttpClientLogicalName { get; } = "HttpEndpointHealthCheckHttpClient";
+
+		internal HttpHealthCheckParameters Parameters { get; } // done internal for unit tests, please don't use it outside of class
```
Comment (neilr81) [inline src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs:21]: "internal only to be used for unit tests" reads better

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.ObjectModel;
+using System.Net;
+using System.Net.Http;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks
+{
+	internal class HttpHealthCheckParameters
+	{
+		public string EndpointName { get; }
```
Comment (neilr81) [inline src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:15]: are these expected to be unique?  Do we have any other identifier?

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+			AdditionalCheck = additionalCheck;
+
+			if (reportData.Length == 0)
+			{
+				ReportData = s_emptyDictionary;
+			}
+			else
+			{
+				// should be replaced by passing enumerable to constructor after we'll drop full framework support
```
Comment (neilr81) [inline src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:58]: we drop

File: src/Extensions/Diagnostics.HealthChecks/Internal/ServiceFabricHealthCheckPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+					Description = CreateDescription(entry)
+				});
+			}
+
+			return Task.CompletedTask;
+		}
+
+		private string CreateDescription(HealthReportEntry entry)
+		{
+			// Healthy reports won't be displayed and most of the checks would be healthy,
```
Comment (neilr81) [inline src/Extensions/Diagnostics.HealthChecks/Internal/ServiceFabricHealthCheckPublisher.cs:53]: will be healthy
creating a detailed description
and avoiding allocations

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs (hunk 1 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs]: Thank you, updated

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs (hunk 3 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs]: Updated, thank you

File: src/Extensions/Diagnostics.HealthChecks/Internal/ServiceFabricHealthCheckPublisher.cs (hunk 4 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/ServiceFabricHealthCheckPublisher.cs]: Fixed, thank you

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs (hunk 2 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:15]: This is the name of the endpoint for current service, they should be unique. I think server creation will fail if you will try to create two endpoints with the same name.
Why would we need other identifiers?

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Omex.Extensions.Abstractions;
+using Microsoft.Omex.Extensions.Abstractions.Activities;
+
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks
+{
+	internal class HttpEndpointHealthCheck : AbstractHealthCheck
```
Comment (dakanev-msft) [inline src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs:15]: >HttpEndpointHealthCheck [](start = 16, length = 23)

Would it be possible to add some logging to be able to troubleshoot if anything goes wrong with the healthcheck logic?

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs (hunk 2 above)
Comment (dakanev-msft) [inline src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:15]: new lines between properties

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+namespace Microsoft.Omex.Extensions.Diagnostics.HealthChecks
+{
+	internal class HttpHealthCheckParameters
+	{
+		public string EndpointName { get; }
+		public Uri RelativeUri { get; }
+		public HttpMethod Method { get; }
+		public HttpStatusCode ExpectedStatus { get; }
+		public string Scheme { get; }
+		public Func<HttpResponseMessage, HealthCheckResult, HealthCheckResult>? AdditionalCheck { get; }
```
Comment (dakanev-msft) [inline src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:20]: >AdditionalCheck [](start = 74, length = 15)

please add some comments about the properties, if possible including examples

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs (hunk 5 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/HttpEndpointHealthCheck.cs:16]: Good point, added  more log calls, thank you

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs (hunk 2 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:15]: added, thanks

File: src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs (hunk 6 above)
Comment (AndreyTretyak) [reply src/Extensions/Diagnostics.HealthChecks/Internal/HttpHealthCheckParameters.cs:25]: I've added a comment with the reference to the public method that has information about all of those fields

File: tests/Extensions/Diagnostics.HealthChecks.UnitTests/HttpEndpointHealthCheckTests.cs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+			Assert.AreEqual(HealthStatus.Unhealthy, result.Status,
+				FormattableString.Invariant($"Should return {HealthStatus.Unhealthy} for wrong status"));
+
+			Assert.AreEqual(contentText, result.Description, "Content should be in the description for unhealthy check");
+			CollectionAssert.AreEquivalent(reportData, result.Data.ToArray(), "Result should propagate reportData");
+		}
+
+		[TestMethod]
+		public async Task CheckHealthAsync_WithAdditionalCheck_ReturnsOverridedResult()
```
Comment (neilr81) [inline tests/Extensions/Diagnostics.HealthChecks.UnitTests/HttpEndpointHealthCheckTests.cs:73]: ReturnsOverridenResult

Comment (AndreyTretyak) [reply tests/Extensions/Diagnostics.HealthChecks.UnitTests/HttpEndpointHealthCheckTests.cs]: Fixed, thank you

**Pull request merged**